metrics/sinks/gcm/gcm.go

```diff
@@ -321,21 +325,43 @@ func CreateGCMSink(uri *url.URL) (core.DataSink, error) {
 		return nil, fmt.Errorf("invalid metrics parameter: %s", metrics)
 	}
 
-	if err := gce_util.EnsureOnGCE(); err != nil {
-		return nil, err
-	}
-
-	// Detect project ID
-	projectId, err := gce.ProjectID()
+	client, err := google.DefaultClient(oauth2.NoContext, gcm.MonitoringScope)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating oauth2 client: %s", err)
 	}
 
-	// Create Google Cloud Monitoring service.
-	client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
 	gcmService, err := gcm.New(client)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating GCM service: %s", err)
+	}
+
+	// Try and get the project ID from an environment variable first
+	var projectId string
+	projectId = os.Getenv(gcpProjectId)
+	if projectId == "" {
+		// If the variable is not set, move on to the default credentials file
+		file := os.Getenv(gcpCredentialEnv)
+		if file != "" {
+			// Attempt to load the config from the credentials file
+			conf, err := ioutil.ReadFile(file)
+			if err != nil {
+				return nil, fmt.Errorf("error reading default credentials file: %s", err)
+			}
+			var gcpConfig struct {
+				ProjectId string `json:"project_id"`
+			}
+			err = json.Unmarshal(conf, &gcpConfig)
+			if err != nil {
+				return nil, fmt.Errorf("error loading default credentials file: %s", err)
```

Comment:
I think that if the credentials are supplied in PK12 format, then `google.DefaultClient(oauth2.NoContext, gcm.MonitoringScope)` will fail earlier in the process? At least from testing it locally, I get:
```
error getting credentials using GOOGLE_APPLICATION_CREDENTIALS environment variable: invalid character '\u0082' after top-level value
```
if I point `GOOGLE_APPLICATION_CREDENTIALS` at a downloaded p12 file. 
But we could still have it fall back on the metadata server if there was a problem unmarshalling the json file for any other reason. 

I'm also thinking it would make sense to move some of this out into some functions in `common/gce/gce.go` to keep everything a bit tidier (`GetProjectIdFromEnv()`, `GetProjectIdFromFile()`, etc). 

Thanks again for taking the time to review this.